### DISCUSSION TITLE
Proposal to add overload operators (e.g. Vector3 / Vector4)

### DISCUSSION THREAD

**Discussion by poirierlouis** [wopss/RED4ext.SDK#104] [Ideas]
Hi,

I'm opening this to discuss more widely about this [PR](https://github.com/maximegmd/CyberEngineTweaks/pull/916).
As @WSSDude said, it shall be better to provide such overload operators right into this SDK, and simply bind them in CET.

There is a concern about performance and how native global functions already implement such arithmetic operations. There is a tiny benchmark I made (see PR) to provide some insights about what is going on. But it probably needs more investigation to be sure.

I'm willing to contribute here, but I'm not aware on how I could write overload operators without messing with generated files.
Here are some points which I guess needs to be clarified:

- [ ] implement overload operators right into SDK?
- [ ] reuse native globals (at the cost of calling functions) OR
- [ ] write raw/custom implementation
- [ ] expand to other operations (dot, cross product, etc.)
- [ ] expand to other types (Quaternion, Matrix, etc.)

**Comment by wopss** [comment]
Hello,

Regarding native and scripted functions, I understand both perspectives. I believe the choice between them depends on the implementer (which might be you in this case), as scripted functions require less effort while creating them in C++ entails more work.

In my view, the results should consistent with the game's functions. Therefore, utilizing natives would be the safest approach and my recommended approach.

> I'm willing to contribute here, but I'm not aware on how I could write overload operators without messing with generated files.

We can do it as we do for [entEntityID](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/Scripting/Natives/entEntityID.hpp) or even [TweakDBID](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/NativeTypes.hpp#L46C8-L46C17).

@psiberx @jackhumbert @jac3km4 what do you guys think?

**Reply by wopss** [reply]
Also, what is your opinion about this?

**Reply by psiberx** [reply]
I'd add our own implementation. Most of math functions are inline anyway. And thanks to many of them being in RTTI for scripts, it's easy to check/grab the exact implementation.

**Reply by poirierlouis** [reply]
Pretty much like @psiberx said. I think its the best to provide the best performance and replicate the same behaviors.

> thanks to many of them being in RTTI for scripts, it's easy to check/grab the exact implementation.

I guess it requires to reverse-engineer the executable to do so?

> We can do it as we do for [entEntityID](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/Scripting/Natives/entEntityID.hpp) or even [TweakDBID](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/NativeTypes.hpp#L46C8-L46C17).

Thanks, I'll check this later, at least for Vector3 / Vector4 to start from something.

**Reply by psiberx** [reply]
> I guess it requires to reverse-engineer the executable to do so?

Yeah. It's easy for RTTI functions and operators.
